xarray/core/duck_array_ops.py

```diff
@@ -53,6 +65,64 @@
 dask_available = module_available("dask")
 
 
+HANDLED_EXTENSION_ARRAY_FUNCTIONS: dict[Callable, Callable] = {}
+
+
+def implements(numpy_function):
+    """Register an __array_function__ implementation for MyArray objects."""
+
+    def decorator(func):
+        HANDLED_EXTENSION_ARRAY_FUNCTIONS[numpy_function] = func
+        return func
+
+    return decorator
+
+
+@implements(np.issubdtype)
+@dispatch
+def __extension_duck_array__issubdtype(
+    extension_array_dtype: T_ExtensionArray, other_dtype: DTypeLikeSave
+) -> bool:
+    return False  # never want a function to think a pandas extension dtype is a subtype of numpy
+
+
+@implements(np.broadcast_to)
+@dispatch
+def __extension_duck_array__broadcast(arr: T_ExtensionArray, shape: tuple):
+    if shape[0] == len(arr) and len(shape) == 1:
+        return arr
+    raise NotImplementedError("Cannot broadcast 1d-only pandas categorical array.")
+
+
+@implements(np.stack)
+@dispatch
+def __extension_duck_array__stack(arr: T_ExtensionArray, axis: int):
+    raise NotImplementedError("Cannot stack 1d-only pandas categorical array.")
+
+
+@implements(np.concatenate)
+@dispatch
```

Comment:
I don't understand why `plum` is needed. We should be using the `__array_function__` system, which you've implemented on `ExtensionDuckArray`. You could move all of these methods to `ExtensionDuckArray` and hardcode `HANDLED_EXTENSION_ARRAY_FUNCTIONS` as a class property.

Reply:
My issue was twofold: one, how can we let people register their own versions of methods for different extension arrays?  Maybe this isn't necessary.  And secondly, the `types` argument, which is meant for dispatching in `__array_function__`, seemed a little too weak for our case.  I couldn't see how to resolve the underlying type of the `ExtensionDuckArray` that was in the 
`types` (see attached).
<img width="377" alt="Screenshot 2024-02-13 at 11 07 00" src="https://github.com/pydata/xarray/assets/43999641/1a010b3d-8d98-4142-9496-434e46f6b77e">
  We could do generic typing plus some very deep properties to recover the generic.  Even so, it wouldn't pick up two different `ExtensionDuckArray`s with different generic types.  Relatedly, I thought we may want to allow for mixed types like `np.where(cond: np.ndarray, x: pd.Categorical, y: MyExtensionArray)`.

Reply:
I don't understand why you need to know the type wrapped by the `ExtensionDuckArray`. If a different version of e.g. `.mean()` is required for a different wrapped type, that should be the job of the duck array class itself to do the dispatching, not xarray.

Reply:
Maybe I'm misunderstanding, but when you say "that should be the job of the duck array class itself to do the dispatching" that sounds like `ExtensionDuckArray` because that _is_ the duck array class.  The reason we need this class is that the pandas extension arrays don't look enough like duck arrays.